PSReadLine/Render.cs

```diff
@@ -340,135 +347,330 @@ void MaybeEmphasize(int i, string currColor)
+                    //
+                    // If the current logical lines are less than the previous, and the cursor is at the beginning of the first line in buffer,
+                    // then it's possible we are facing this special case and thus would need to do additional checks later.
+
+                    minLineLength = renderLines.Length;
+                    if (cursorX == Options.ContinuationPrompt.Length && cursorY == 0)
                     {
-                        renderErrorPrompt = true;
-                        _console.CursorLeft -= promptBufferCells;
+                        // Number of physical lines before counting the first line in buffer.
+                        linesToCheck = 0 - _initialY;
                     }
-                    else
+                }
+
+                // Find the first logical line that was changed.
+                for (; logicalLine < minLineLength; logicalLine++)
+                {
+                    // Found the first different logical line? Break out the loop.
+                    if (renderLines[logicalLine].line != previousRenderLines[logicalLine].line) { break; }
+
+                    int count = PhysicalLineCount(renderLines[logicalLine].columns, logicalLine == 0, out _);
+                    physicalLine += count;
+
+                    if (physicalLine == linesToCheck && previousRenderLines[logicalLine + 1].columns == Options.ContinuationPrompt.Length)
                     {
-                        // The 'CursorLeft' could be less than error-prompt-cell-length in one of the following 3 cases:
-                        //   1. console buffer was resized, which causes the initial cursor to appear on the next line;
-                        //   2. prompt string gets longer (e.g. by 'cd' into nested folders), which causes the line to be wrapped to the next line;
-                        //   3. the prompt function was changed, which causes the new prompt string is shorter than the error prompt.
-                        // Here, we always assume it's the case 1 or 2, and wrap back to the previous line to change the error prompt color.
-                        // In case of case 3, the rendering would be off, but it's more of a user error because the prompt is changed without
-                        // updating 'PromptText' with 'Set-PSReadLineOption'.
-
-                        int diffs = promptBufferCells - _console.CursorLeft;
-                        int newX = bufferWidth - diffs % bufferWidth;
-                        int newY = _initialY - diffs / bufferWidth - 1;
-
-                        // newY could be less than 0 if 'PromptText' is manually set to be a long string.
-                        if (newY >= 0)
+                        // Additional check for the special case mentioned above: the cursor is supposed to be moved to the previous line.
+                        if (ConvertOffsetToPoint(_current).Y == -1)
```

Comment:
This issue has been addressed, the logic is changed to cover the more general scenarios where the cursor is supposed to be moved to a position that is off the buffer after the editing.
In the following screenshot, the `Hello world` line and the following 3 empty lines are moved off the buffer, and then <kbd>Alt+4</kbd><kbd>Backspace</kbd> works as expected.

![fix1](https://user-images.githubusercontent.com/127450/64053712-61260700-cb38-11e9-85c8-4f64f91abe38.gif)

However, as you might noticed, I removed the last empty line before moving the cursor to `<0, 0>` position. This is because if I don't remove the last line, <kbd>Alt+4</kbd> scroll up the buffer to show the `digit-argument:` line, and will move the cursor one line above. In this case, the cursor will be moved to `<0, -1>`, and thus causes an exception.

This is not a problem introduced by this change. See the following screenshot with the `beta.4` PSReadLine:

![bk1](https://user-images.githubusercontent.com/127450/64054135-8582e300-cb3a-11e9-8a99-8f58a9e3b158.gif)